### PR TITLE
[Bugfix]: 'forbidden_chars' - the condition added for uploading via Drop or UploadPicker. #43495

### LINKED ISSUE #43423: [Bug]: 'forbidden_chars' Not Enforced During File Uploads
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

The parameter 'forbidden_chars,' introduced in Nextcloud 28 for setting prohibited characters, applies restrictions only when creating or renaming files and folders through the WebUI. This restriction does not extend to file uploads.

Could you please update the functionality to ensure that this parameter is also applied during file uploads?

The restriction does not apply to file uploads through the WebUI drag and drop or the "Upload files" using the +New button. 

### Steps to reproduce

1. + New
2. Upload files
3. Test#.txt

### Expected behavior

Thest "Test#.txt" should also be declined because of the restricted character like it is already done for folders

### Installation method

None

### Nextcloud Server version

28

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.2

### Web server

Nginx

### Database engine version

MySQL

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

_No response_

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by arublov**
* Resolves: #43423  

## Summary
1. Added the condition for uploading files via drop.
2. Added the forbiddenCharacters parameter for UploadPicker (need to merge [this pull request](https://github.com/nextcloud-libraries/nextcloud-upload/pull/1086)).

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by arublov** [comment]
@szaimen I apologize for the confusion I caused. I got a bit tangled in my environment, had to optimize it a bit, and as a result, it led to some spam. I'll try not to let this happen again in the future :)

Thanks for your help and comment.

File: apps/files/src/services/DropService.ts [hunk 1]
```diff
@@ -67,11 +70,18 @@ export const handleDrop = async (data: DataTransfer): Promise<Upload[]> => {
...
-	try {
-		return await uploader.upload(`${path}${file.name}`, file)
-	} catch (e) {
-		showError(t('files', 'Uploading "{filename}" failed', { filename: file.name }))
-		throw e
+	const forbidden = forbiddenCharacters.split('')
+	let forbiddenChar
+
+	if(forbiddenChar = forbidden.find(char => file.name.includes(char))){
+		showError(t('files', `"${forbiddenChar}" is not allowed inside a file name.`));
```
Comment (susnux) [inline apps/files/src/services/DropService.ts:77]: This does not for translations. Instead:
```suggestion
		showError(t('files', '"{forbiddenChar}" is not allowed inside a file name.', { forbiddenChar }))
```

File: apps/files/src/services/DropService.ts [hunk 2]
```diff
@@ -67,11 +70,18 @@ export const handleDrop = async (data: DataTransfer): Promise<Upload[]> => {
...
 
-	try {
-		return await uploader.upload(`${path}${file.name}`, file)
-	} catch (e) {
-		showError(t('files', 'Uploading "{filename}" failed', { filename: file.name }))
-		throw e
+	const forbidden = forbiddenCharacters.split('')
+	let forbiddenChar
+
+	if(forbiddenChar = forbidden.find(char => file.name.includes(char))){
```
Comment (susnux) [inline apps/files/src/services/DropService.ts:76]: I wonder ESLint is not complaining, spaces missing here for our codestyle

```suggestion
	if (forbiddenChar = forbidden.find(char => file.name.includes(char))) {
```

File: apps/files/src/services/DropService.ts [hunk 3]
```diff
@@ -67,11 +70,18 @@ export const handleDrop = async (data: DataTransfer): Promise<Upload[]> => {
...
+		showError(t('files', `"${forbiddenChar}" is not allowed inside a file name.`));
+	}else{
```
Comment (susnux) [inline apps/files/src/services/DropService.ts:78]: same here about our code style
```suggestion
	} else {
```

File: apps/files/src/services/DropService.ts (hunk 2 above)
Comment (susnux) [inline apps/files/src/services/DropService.ts:76]: We should not assign variables within the if condition.

```suggestion
	const forbiddenChar = forbidden.find(char => file.name.includes(char))

	if(forbiddenChar !== undefined){
```

**Review by susnux — CHANGES_REQUESTED** [review]
Some comments about the codestyle (really not sure why ESLint does not complain?)

File: apps/files/src/services/DropService.ts (hunk 1 above)
Comment (arublov) [reply apps/files/src/services/DropService.ts:77]: @susnux yes, true, my mistake, thanks

**Comment by arublov** [comment]
> Some comments about the codestyle (really not sure why ESLint does not complain?)

@susnux I fixed my code according to your comments. Thank you.
Do I need to do something with ESLint? Maybe something on my side?

File: apps/files/src/services/DropService.ts [hunk 4]
```diff
@@ -29,9 +29,12 @@ import { getUploader } from '@nextcloud/upload'
 import { joinPaths } from '@nextcloud/paths'
 import { showError } from '@nextcloud/dialogs'
 import { translate as t } from '@nextcloud/l10n'
+import { loadState } from '@nextcloud/initial-state'
 
 import logger from '../logger.js'
 
+const forbiddenCharacters = loadState('files', 'forbiddenCharacters', '') as string
```
Comment (susnux) [inline apps/files/src/services/DropService.ts:36]:
```suggestion
const forbiddenCharacters = loadState<string>('files', 'forbiddenCharacters', '')
```

No explicit cast needed

File: apps/files/src/services/DropService.ts [hunk 5]
```diff
@@ -67,11 +70,18 @@ export const handleDrop = async (data: DataTransfer): Promise<Upload[]> => {
...
 
-	try {
-		return await uploader.upload(`${path}${file.name}`, file)
-	} catch (e) {
-		showError(t('files', 'Uploading "{filename}" failed', { filename: file.name }))
-		throw e
+	const forbidden = forbiddenCharacters.split('')
+	const forbiddenChar = forbidden.find(char => file.name.includes(char))
+
+	if(forbiddenChar !== undefined){
```
Comment (susnux) [inline apps/files/src/services/DropService.ts:76]:
```suggestion
	if (forbiddenChar !== undefined) {
```

**Review by susnux — APPROVED** [review]
besides that two nits it looks good to me

File: apps/files/src/views/FilesList.vue [hunk 6]
```diff
@@ -402,6 +404,10 @@ export default defineComponent({
 			return isSharingEnabled
 				&& this.currentFolder && (this.currentFolder.permissions & Permission.SHARE) !== 0
 		},
+
+		forbiddenCharacters(): string {
+			return forbiddenCharacters
+		},
```
Comment (skjnldsv) [inline apps/files/src/views/FilesList.vue:410]: This should be in `data`, and not a computed :)

Comment (susnux) [reply apps/files/src/views/FilesList.vue:410]: No it should not be in data as that would uselessly make it reactive. Especially as it is const.
I would say either computed or static instance attribute  using `setup`.

**Comment by joshtrichards** [comment]
Best as I can tell this was superseded by nextcloud-libraries/nextcloud-upload#1132 (fka: nextcloud-libraries/nextcloud-upload#1086) and 079d026d39565e7b5547741bd62bd69082347ad6

**Pull request closed without merging**